### PR TITLE
New: Check Node version before run

### PR DESCRIPTION
**Pull request by stramel**
Prints a message if the user doesn't have the minimum version of Node to run this tool.

### CODE REVIEW

**Comment by stramel** [comment]
@FredKSchott @usergenic @rictic Rebased onto the latest changes in master

File: bin/modulizer.js [hunk 1]
```diff
@@ -18,12 +18,34 @@
...
+ * asynchronously after this version check has been performed.
+ */
+
+var semver = require('semver');
+var version = require('../package.json').engines.node;
+
+// Exit early if the user's node version is too low.
+if (!semver.satisfies(process.version, version)) {
+  // Strip version range characters leaving the raw semantic version for output
+  var rawVersion = version.replace(/[^\d\.]*/, '');
```
Comment (FredKSchott) [inline bin/modulizer.js]: The engines field might change to be something with a more complex format. It's fine to just report that string, something like:

`'Polymer Modulizer requires Node version ' + engines.node + '. You have...`

File: bin/modulizer.js [hunk 2]
```diff
@@ -18,12 +18,34 @@
...
+  console.log(
+      'Polymer Modulizer requires at least Node v' + rawVersion + '. ' +
+      'You have ' + process.version + '.\n' +
+      'See polymer-project.org/2.0/docs/tools/node-support ' +
```
Comment (FredKSchott) [inline bin/modulizer.js]: https://www.polymer-project.org/2.0/docs/tools/node-support

File: package.json [hunk 3]
```diff
@@ -26,7 +26,7 @@
     "prepare": "npm run build"
   },
   "engines": {
-    "node": ">=7.6.0"
+    "node": ">=8.9.0"
```
Comment (FredKSchott) [inline package.json]: lets make this `>=8.0.0`, the only reason we specifically targetting 7.6.0 was because that was when async/await was introduced.

File: bin/modulizer.js (hunk 1 above)
Comment (FredKSchott) [reply bin/modulizer.js]: to clarify: remove `rawVersion` formatting and just log the raw `engines.node` string

File: package.json (hunk 3 above)
Comment (stramel) [reply package.json]: Okay, I will change that. I wasn't sure if I should use 8 or the first LTS version.

File: bin/modulizer.js (hunk 1 above)
Comment (stramel) [reply bin/modulizer.js]: Sorry, I misread your comment. Updating. 👍

File: bin/modulizer.js [hunk 4]
```diff
@@ -18,12 +18,32 @@
...
 process.title = 'modulizer';
 
+/*
+ * NOTE: The contents of this file should work on as many version of Node.js
+ * as possible. This means it *can not* use any >ES5 syntax and features.
+ * Other files, which may use >=ES2015 syntax, should only be loaded
+ * asynchronously after this version check has been performed.
+ */
+
+var semver = require('semver');
```
Comment (justinfagnani) [inline bin/modulizer.js]: I'm fine with letting `let` just fail on super old node. We have the engine field in package.json for that anyway

File: bin/modulizer.js [hunk 5]
```diff
@@ -18,12 +18,32 @@
...
+      'You have ' + process.version + '.\n' +
+      'See https://www.polymer-project.org/2.0/docs/tools/node-support ' +
+      'for details.');
+  process.exit(1);
+}
+
 require('source-map-support').install();
 
-const {run} = require('../lib/cli/index');
+var cli = require('../lib/cli/index');
```
Comment (justinfagnani) [inline bin/modulizer.js]: move back to const

File: bin/modulizer.js (hunk 4 above)
Comment (stramel) [reply bin/modulizer.js]: @justinfagnani What is the benefit of `const`/`let` over `var` here? Seems like it would be better to just use `var` in here to provide a better message to the user. Especially since that is the point of this PR is to provide a nice message to the user of older Node versions. However it appears that `const` is supported all the way back to `0.10.x` and should suffice for this. 

http://node.green/#ES2015-bindings-const-basic-support

Comment (FredKSchott) [reply bin/modulizer.js]: If the entire point of this file is "give users outside of our support policy a helpful message" Then var > const makes sense here. Yes, var-only Node versions are super out of date, but that's the whole reason we have this file so why wouldn't we?

This is a small point so I'm fine merging with either, but var makes more sense to me.

Comment (justinfagnani) [reply bin/modulizer.js]: The benefit is consistency, and afaict there aren't any downsides as the engines field takes care of older node versions.

In fact, I wonder why we would add our own custom logic when this is exactly what `engines` is for?

Comment (FredKSchott) [reply bin/modulizer.js]: Agreed that's exactly what `engines` should be for, but `engines` isn't enforced by npm. I'm not even sure if you get a warning anymore

Comment (stramel) [reply bin/modulizer.js]: @justinfagnani @FredKSchott 

So I decided to give this a shot just to see what would happen trying to run this on an older version. 

**TL;DR** 
- Yarn enforces the engine version while NPM does not.
- A user can bypass the engine check with `--ignore-engines` or `npm i -g polymer-modulizer`

<details><summary>Attempting with v0.10.48</summary>

```sh
chrx@chrx:~$ nvm install 0.10
Downloading and installing node v0.10.48...
Now using node v0.10.48 (npm v2.15.1)
chrx@chrx:~$ node -v
v0.10.48
chrx@chrx:~$ npm i -g yarn
npm WARN engine yarn@1.3.2: wanted: {"node":">=4.0.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
yarn@1.3.2 /home/chrx/.nvm/v0.10.48/lib/node_modules/yarn
chrx@chrx:~$ yarn global add polymer-modulizer
Node version 0.10.48 is not supported, please use Node.js 4.0 or higher.
chrx@chrx:~$ cd Projects/Github/polymer-modulizer
chrx@chrx:~/Projects/Github/polymer-modulizer$ npm i -g .
npm WARN engine polymer-modulizer@0.2.1: wanted: {"node":">=8.0.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine jscodeshift@0.3.32: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine command-line-usage@4.0.1: wanted: {"node":">=4.0.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine chalk@2.3.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine polymer-analyzer@2.6.0: wanted: {"node":">=6"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine esprima@3.1.3: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine supports-color@4.5.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine ansi-styles@3.2.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine ansi-escape-sequences@4.0.0: wanted: {"node":">=4.0.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine table-layout@0.4.2: wanted: {"node":">=4.0.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine array-back@2.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine run-async@2.3.0: wanted: {"node":">=0.12.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine string-width@2.1.1: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine ansi-escapes@3.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine cli-cursor@2.1.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine strip-ansi@4.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine external-editor@2.0.5: wanted: {"node":">=0.12"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine figures@2.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine wordwrapjs@3.0.0: wanted: {"node":">=4.0.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine is-fullwidth-code-point@2.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine ansi-regex@3.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine restore-cursor@2.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine array-back@2.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine find-replace@1.0.3: wanted: {"node":">=4.0.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine esprima@4.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine onetime@2.0.1: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine array-back@1.0.4: wanted: {"node":">=0.12.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine mimic-fn@1.1.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN deprecated babel-preset-es2015@6.24.1: 🙌  Thanks for using Babel: we recommend using babel-preset-env now: please read babeljs.io/env to update! 
npm WARN engine escodegen@1.9.0: wanted: {"node":">=0.12.0"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN engine strip-indent@2.0.0: wanted: {"node":">=4"} (current: {"node":"0.10.48","npm":"2.15.1"})
npm WARN deprecated minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
/home/chrx/.nvm/v0.10.48/bin/modulizer -> /home/chrx/.nvm/v0.10.48/lib/node_modules/polymer-modulizer/bin/modulizer.js
polymer-modulizer@0.2.1 /home/chrx/.nvm/v0.10.48/lib/node_modules/polymer-modulizer
├── @types/semver@5.4.0
├── @types/chai@3.5.2
├── semver@5.4.1
├── tslib@1.8.0
├── @types/esprima@2.1.35 (@types/estree@0.0.38)
├── ast-types@0.9.14
├── mkdirp@0.5.1 (minimist@0.0.8)
├── mz@2.7.0 (object-assign@4.1.1, thenify-all@1.6.0, any-promise@1.3.0)
├── esprima@3.1.3
├── chalk@2.3.0 (escape-string-regexp@1.0.5, supports-color@4.5.0, ansi-styles@3.2.0)
├── command-line-usage@4.0.1 (typical@2.6.1, ansi-escape-sequences@4.0.0, array-back@2.0.0, table-layout@0.4.2)
├── command-line-args@4.0.7 (typical@2.6.1, array-back@2.0.0, find-replace@1.0.3)
├── rimraf@2.6.2 (glob@7.1.2)
├── @types/rimraf@2.0.2 (@types/glob@5.0.33, @types/node@8.0.52)
├── @types/mkdirp@0.5.1 (@types/node@8.0.52)
├── @types/mz@0.0.31 (@types/node@8.0.52)
├── @types/inquirer@0.0.35 (@types/rx@4.1.1, @types/through@0.0.29)
├── chai@3.5.0 (assertion-error@1.0.2, type-detect@1.0.0, deep-eql@0.1.3)
├── source-map-support@0.4.18 (source-map@0.5.7)
├── polymer-workspaces@1.1.0 (github@11.0.0)
├── inquirer@3.3.0 (ansi-escapes@3.0.0, cli-width@2.2.0, through@2.3.8, figures@2.0.0, string-width@2.1.1, strip-ansi@4.0.0, rx-lite-aggregates@4.0.8, run-async@2.3.0, mute-stream@0.0.7, cli-cursor@2.1.0, rx-lite@4.0.8, external-editor@2.0.5, lodash@4.17.4)
├── polymer-analyzer@2.6.0 (strip-indent@2.0.0, @types/estraverse@0.0.6, @types/chai-subset@1.3.1, @types/cssbeautify@0.3.1, @types/clone@0.1.30, @types/chalk@0.4.31, @types/doctrine@0.0.1, @types/escodegen@0.0.2, @types/parse5@2.2.34, @types/estree@0.0.37, clone@2.1.1, estraverse@4.2.0, dom5@2.3.0, cssbeautify@0.3.1, jsonschema@1.2.0, chalk@1.1.3, @types/node@6.0.91, doctrine@2.0.0, parse5@2.2.3, shady-css-parser@0.1.0, espree@3.5.2, escodegen@1.9.0)
├── recast@0.12.9 (private@0.1.8, ast-types@0.10.1, esprima@4.0.0, source-map@0.6.1, core-js@2.5.1)
├── jscodeshift@0.3.32 (node-dir@0.1.8, colors@1.1.2, write-file-atomic@1.3.4, temp@0.8.3, async@1.5.2, nomnom@1.8.1, babylon@6.18.0, micromatch@2.3.11, flow-parser@0.59.0, lodash@4.17.4, babel-register@6.26.0, babel-plugin-transform-flow-strip-types@6.22.0, babel-preset-stage-1@6.24.1, babel-core@5.8.38, babel-preset-es2015@6.24.1)
└── ix@2.3.1
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ modulizer --in ../marked-element
Polymer Modulizer requires Node version >=8.0.0. You have v0.10.48.
See https://www.polymer-project.org/2.0/docs/tools/node-support for details.
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ 
```

NOTE: `yarn` itself requires v4.0 to run.
NOTE: `npm i -g polymer-modulizer` will still run if it doesn't meet engine requirements but does output warnings for every missed engine version.
NOTE: `const` did not work, resulting in this error.
```sh
const semver = require('semver');
^^^^^
SyntaxError: Use of const in strict mode.
    at Module._compile (module.js:439:25)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:945:3
```
</details>

<details><summary>Attempting with v4.8.0 (minimum for Yarn)</summary>

```sh
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ nvm install 4.8.0
Downloading and installing node v4.8.0...
Now using node v4.8.0 (npm v2.15.11)
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ node -v
v4.8.0
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ npm i -g yarn
yarn@1.3.2 /home/chrx/.nvm/versions/node/v4.0.0/lib/node_modules/yarn
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ yarn global add polymer-modulizer
yarn global v1.3.2
[1/4] Resolving packages...
warning polymer-modulizer > bower@1.8.2: ...psst! Your project can stop working at any moment because its dependencies can change. Prevent this by migrating to Yarn: https://bower.io/blog/2017/how-to-migrate-away-from-bower/
warning polymer-modulizer > jscodeshift > babel-preset-es2015@6.24.1: 🙌  Thanks for using Babel: we recommend using babel-preset-env now: please read babeljs.io/env to update! 
warning polymer-modulizer > jscodeshift > babel-core > minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
[2/4] Fetching packages...
error polymer-analyzer@2.6.0: The engine "node" is incompatible with this module. Expected version ">=6".
error Found incompatible module
info Visit https://yarnpkg.com/en/docs/cli/global for documentation about this command.
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ npm i -g .
npm WARN engine polymer-modulizer@0.2.1: wanted: {"node":">=8.0.0"} (current: {"node":"4.8.0","npm":"2.15.11"})
npm WARN engine polymer-analyzer@2.6.0: wanted: {"node":">=6"} (current: {"node":"4.8.0","npm":"2.15.11"})
npm WARN deprecated babel-preset-es2015@6.24.1: 🙌  Thanks for using Babel: we recommend using babel-preset-env now: please read babeljs.io/env to update! 
npm WARN deprecated minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
/home/chrx/.nvm/versions/node/v4.8.0/bin/modulizer -> /home/chrx/.nvm/versions/node/v4.8.0/lib/node_modules/polymer-modulizer/bin/modulizer.js
polymer-modulizer@0.2.1 /home/chrx/.nvm/versions/node/v4.8.0/lib/node_modules/polymer-modulizer
├── @types/semver@5.4.0
├── @types/chai@3.5.2
├── semver@5.4.1
├── esprima@3.1.3
├── tslib@1.8.0
├── @types/esprima@2.1.35 (@types/estree@0.0.38)
├── ast-types@0.9.14
├── @types/mkdirp@0.5.1 (@types/node@8.0.52)
├── @types/mz@0.0.31 (@types/node@8.0.52)
├── @types/rimraf@2.0.2 (@types/node@8.0.52, @types/glob@5.0.33)
├── @types/inquirer@0.0.35 (@types/rx@4.1.1, @types/through@0.0.29)
├── source-map-support@0.4.18 (source-map@0.5.7)
├── mkdirp@0.5.1 (minimist@0.0.8)
├── mz@2.7.0 (object-assign@4.1.1, thenify-all@1.6.0, any-promise@1.3.0)
├── chalk@2.3.0 (escape-string-regexp@1.0.5, supports-color@4.5.0, ansi-styles@3.2.0)
├── rimraf@2.6.2 (glob@7.1.2)
├── chai@3.5.0 (assertion-error@1.0.2, type-detect@1.0.0, deep-eql@0.1.3)
├── command-line-usage@4.0.1 (ansi-escape-sequences@4.0.0, typical@2.6.1, array-back@2.0.0, table-layout@0.4.2)
├── command-line-args@4.0.7 (typical@2.6.1, array-back@2.0.0, find-replace@1.0.3)
├── polymer-workspaces@1.1.0 (github@11.0.0)
├── polymer-analyzer@2.6.0 (strip-indent@2.0.0, @types/estree@0.0.37, @types/escodegen@0.0.2, @types/cssbeautify@0.3.1, @types/estraverse@0.0.6, @types/parse5@2.2.34, @types/clone@0.1.30, @types/chai-subset@1.3.1, @types/chalk@0.4.31, @types/doctrine@0.0.1, estraverse@4.2.0, clone@2.1.1, @types/node@6.0.91, cssbeautify@0.3.1, dom5@2.3.0, jsonschema@1.2.0, parse5@2.2.3, chalk@1.1.3, doctrine@2.0.0, escodegen@1.9.0, shady-css-parser@0.1.0, espree@3.5.2)
├── inquirer@3.3.0 (ansi-escapes@3.0.0, rx-lite-aggregates@4.0.8, cli-width@2.2.0, rx-lite@4.0.8, through@2.3.8, mute-stream@0.0.7, figures@2.0.0, strip-ansi@4.0.0, string-width@2.1.1, run-async@2.3.0, cli-cursor@2.1.0, external-editor@2.0.5, lodash@4.17.4)
├── recast@0.12.9 (private@0.1.8, esprima@4.0.0, ast-types@0.10.1, source-map@0.6.1, core-js@2.5.1)
├── ix@2.3.1
└── jscodeshift@0.3.32 (node-dir@0.1.8, async@1.5.2, babylon@6.18.0, colors@1.1.2, write-file-atomic@1.3.4, flow-parser@0.59.0, temp@0.8.3, nomnom@1.8.1, micromatch@2.3.11, lodash@4.17.4, babel-core@5.8.38, babel-register@6.26.0, babel-plugin-transform-flow-strip-types@6.22.0, babel-preset-stage-1@6.24.1, babel-preset-es2015@6.24.1)
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ modulizer --in ../marked-element
Polymer Modulizer requires Node version >=8.0.0. You have v4.8.0.
See https://www.polymer-project.org/2.0/docs/tools/node-support for details.
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ 
```

NOTE: `yarn` prevents `polymer-modulizer` from being added due to engine restrictions.
NOTE: `const` properly works in `v4.8.0`
</details>

<details><summary>Attempting with v4.8.0, Yarn w/ `--ignore-engines` flag</summary>

```sh
chrx@chrx:~/Projects/GitHub/polymer-modulizer$ yarn global add polymer-modulizer --ignore-engines
yarn global v1.3.2
[1/4] Resolving packages...
warning polymer-modulizer > bower@1.8.2: ...psst! Your project can stop working at any moment because its dependencies can change. Prevent this by migrating to Yarn: https://bower.io/blog/2017/how-to-migrate-away-from-bower/
warning polymer-modulizer > jscodeshift > babel-preset-es2015@6.24.1: 🙌  Thanks for using Babel: we recommend using babel-preset-env now: please read babeljs.io/env to update! 
warning polymer-modulizer > jscodeshift > babel-core > minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Installed "polymer-modulizer@0.2.1" with binaries:
      - modulizer
Done in 132.14s.
```

Comment (FredKSchott) [reply bin/modulizer.js]: Good to know, thanks for looking @stramel 

@justinfagnani can you change status to approve so we can merge this? I'm fine with as written (`const`) but don't want to merge while you're ❌

Comment (stramel) [reply bin/modulizer.js]: @fredkschott just wanted to make sure you saw that `const` doesn't work on 0.10

**Comment by stramel** [comment]
@justinfagnani Is there anything else you would like addressed here?

**Comment by justinfagnani** [comment]
I guess this is LGTM, I'm just kind of sad that the engines field doesn't take care of this.

**Comment by stramel** [comment]
Ping @fredkschott
